File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        def dfs(prev: ListNode, node: ListNode | None) -> int:
            if node is None:
                return 1
            
            i = dfs(node, node.next)
            if i == n:
                # prev -> node -> node.next
                # => prev -> node.next
                prev.next = node.next

            return i + 1
        
        prev = ListNode()
        prev.next = head
        dfs(prev, head)
        return prev.next

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode[]} lists
     * @return {ListNode}
     */
    mergeKLists(lists) {
        /** @type {ListNode | null} */
        let merged = null;
        let head = null;

        while (true) {
            let min_i = null;
            for (let i = 0; i < lists.length; i++) {
                if (lists[i] == null) continue;
                if (min_i == null) {
                    min_i = i;
                    continue;
                }
                if (lists[i].val < lists[min_i].val) {
                    min_i = i;
                }
            }

            if (min_i == null) break;

            // merged を更新
            if (merged == null) {
                merged = new ListNode(lists[min_i].val)
                head = merged
            } else {
                merged.next = new ListNode(lists[min_i].val);
                merged = merged.next
            }

            // ポインタを更新
            lists[min_i] = lists[min_i].next;
        }

        return head;
    }
}
